node/primitives/src/approval.rs

```diff
@@ -16,190 +16,507 @@
-	/// The validator index.
-	pub validator: ValidatorIndex,
-	/// The signature by the validator.
-	pub signature: ValidatorSignature,
+		None
+	}
 }
 
-/// Metadata about a block which is now live in the approval protocol.
-#[derive(Debug)]
-pub struct BlockApprovalMeta {
-	/// The hash of the block.
-	pub hash: Hash,
-	/// The number of the block.
-	pub number: BlockNumber,
-	/// The hash of the parent block.
-	pub parent_hash: Hash,
-	/// The candidates included by the block.
-	/// Note that these are not the same as the candidates that appear within the block body.
-	pub candidates: Vec<CandidateHash>,
-	/// The consensus slot of the block.
-	pub slot: Slot,
-	/// The session of the block.
-	pub session: SessionIndex,
-}
+/// A list of primitives introduced by v2.
+pub mod v2 {
+	use parity_scale_codec::{Decode, Encode};
+	pub use sp_consensus_babe::{
+		Randomness, Slot, VrfOutput, VrfProof, VrfSignature, VrfTranscript,
+	};
+	use std::ops::BitOr;
 
-/// Errors that can occur during the approvals protocol.
-#[derive(Debug, thiserror::Error)]
-#[allow(missing_docs)]
-pub enum ApprovalError {
-	#[error("Schnorrkel signature error")]
-	SchnorrkelSignature(schnorrkel::errors::SignatureError),
-	#[error("Authority index {0} out of bounds")]
-	AuthorityOutOfBounds(usize),
-}
+	use bitvec::{prelude::Lsb0, vec::BitVec};
+	use polkadot_primitives::{CandidateIndex, CoreIndex, Hash, ValidatorIndex};
 
-/// An unsafe VRF output. Provide BABE Epoch info to create a `RelayVRFStory`.
-pub struct UnsafeVRFOutput {
-	vrf_output: VrfOutput,
-	slot: Slot,
-	authority_index: u32,
-}
+	/// A static context associated with producing randomness for a core.
+	pub const CORE_RANDOMNESS_CONTEXT: &[u8] = b"A&V CORE v2";
+	/// A static context associated with producing randomness for v2 multi-core assignments.
+	pub const ASSIGNED_CORE_CONTEXT: &[u8] = b"A&V ASSIGNED v2";
+	/// A static context used for all relay-vrf-modulo VRFs for v2 multi-core assignments.
+	pub const RELAY_VRF_MODULO_CONTEXT: &[u8] = b"A&V MOD v2";
+	/// A read-only bitvec wrapper
+	#[derive(Clone, Debug, Encode, Decode, Hash, PartialEq, Eq)]
+	pub struct Bitfield<T>(BitVec<u8, bitvec::order::Lsb0>, std::marker::PhantomData<T>);
+
+	/// A `read-only`, `non-zero` bitfield.
+	/// Each 1 bit identifies a candidate by the bitfield bit index.
+	pub type CandidateBitfield = Bitfield<CandidateIndex>;
+	/// A bitfield of core assignments.
+	pub type CoreBitfield = Bitfield<CoreIndex>;
+
+	/// Errors that can occur when creating and manipulating bitfields.
+	#[derive(Debug)]
+	pub enum BitfieldError {
+		/// All bits are zero.
+		NullAssignment,
+	}
+
+	/// A bit index in `Bitfield`.
+	#[cfg_attr(test, derive(PartialEq, Clone))]
+	pub struct BitIndex(pub usize);
+
+	/// Helper trait to convert primitives to `BitIndex`.
+	pub trait AsBitIndex {
+		/// Returns the index of the corresponding bit in `Bitfield`.
+		fn as_bit_index(&self) -> BitIndex;
+	}
+
+	impl<T> Bitfield<T> {
```

Comment:
It doesn't make much sense for this to be a generic type. Why is it?

Reply:
It's only needed such that CoreBitfield and CandidateBitfield are different types and cannot be mixed up.